src/main/java/org/terasology/logic/health/HealthClientSystem.java

```diff
@@ -57,51 +62,77 @@ public void initialise() {
+                damageDirection.setTop(true);
+                delayManager.addDelayedAction(entity, REMOVE_TOP_ACTION, DAMAGE_OVERLAY_DELAY_MS);
+            } else if (direction <= -45.0 && direction > -135.0) {
+                damageDirection.setRight(true);
+                delayManager.addDelayedAction(entity, REMOVE_RIGHT_ACTION, DAMAGE_OVERLAY_DELAY_MS);
+            } else if (direction <= 135.0 && direction > 45.0) {
+                damageDirection.setLeft(true);
+                delayManager.addDelayedAction(entity, REMOVE_LEFT_ACTION, DAMAGE_OVERLAY_DELAY_MS);
+            } else {
+                damageDirection.setBottom(true);
+                delayManager.addDelayedAction(entity, REMOVE_BOTTOM_ACTION, DAMAGE_OVERLAY_DELAY_MS);
+            }
+        }
 
-            if (!overlayDisplaying) {
-                if (angleDegrees <= 45.0 && angleDegrees > -45.0) {
-                    damageDirection.setTop(true);
-                } else if (angleDegrees <= -45.0 && angleDegrees > -135.0) {
-                    damageDirection.setRight(true);
-                } else if (angleDegrees <= 135.0 && angleDegrees > 45.0) {
-                    damageDirection.setLeft(true);
-                } else {
-                    damageDirection.setBottom(true);
-                }
+        // Is it major damage?
+        HealthComponent healthComponent = entity.getComponent(HealthComponent.class);
+        int amount = event.getDamageAmount();
+        int current = healthComponent.currentHealth;
+        if (current > 0 && amount > 0) {
+            float percent = (float) amount / current;
+            if (percent >= DAMAGE_OVERLAY_REQUIRED_PERCENT && !overlayDisplaying) {
                 overlayDisplaying = true;
-                delayManager.addDelayedAction(entity, REMOVE_DAMAGED_OVERLAY_ACTION, DAMAGED_OVERLAY_DELAY_MS);
+                nuiManager.addOverlay(DAMAGE_OVERLAY, MajorDamageOverlay.class);
+                delayManager.addDelayedAction(entity, REMOVE_DAMAGE_OVERLAY_ACTION, DAMAGE_OVERLAY_DELAY_MS);
             }
         }
     }
 
-    private double rad2Deg(double rad) {
-        return rad / Math.PI * 180.0;
+    private double determineDamageDirection(EntityRef instigator, LocationComponent locationComponent) {
+        LocationComponent instigatorLocation = instigator.getComponent(LocationComponent.class);
+        Vector3f loc = new Vector3f();
+        loc = locationComponent.getWorldPosition(loc);
+        Vector3f instLoc = new Vector3f();
+        instLoc = instigatorLocation.getWorldPosition(instLoc);
+        Vector3f locDiff = instLoc.sub(loc);
+        locDiff.normalize();
+
+        Vector3f worldFacing = new Vector3f();
+        // facing x and z are "how much" of that direction we are facing
+        // e.g. (0.0, 1.0) means that going forward increases world z position without increasing x position
+        worldFacing = locationComponent.getWorldDirection(worldFacing);
+        worldFacing.normalize();
```

Comment:
This is likely a matter of taste - you _could_ combine this into a single line:

```suggestion
        // facing x and z are "how much" of that direction we are facing
        // e.g. (0.0, 1.0) means that going forward increases world z position without increasing x position
		Vector3f worldFacing = locationComponent.getWorldDirection(new Vector3f()).normalize();
```